Extract SSM store, add environment + chained stores

glb opened the pull request:
- SSM parameter store is now in its own class.

- Added an environment parameter store that reads from the OS
  environment, optionally with a prefix.

- Added a chained store (now the default) that will first look
  at the environment and then use SSM as a backup for parameters
  that are not found.

Fixes alexcasalboni/ssm-cache-python#3.

glb commented:
Note: I am not in any way expecting this to sail through, opened the PR to get your comments.

With this change the implementation is now more of a generic parameter store that happens to have SSM as a backing store, so I did some non-backwards-compatible naming changes to make things more generic.

I expect you'll want me to break the change up into multiple commits to clarify the incremental changes; there's also a possibility that I missed something when comparing with other branches that are in flight, so I'd like to serialize a bit and rebase once / if those are merged.

Thoughts?

alexcasalboni commented on ssm_cache/ssm.py:
```diff
@@ -0,0 +1,89 @@
+        return cls._CLIENT
+
+    @classmethod
+    def _batch(cls, iterable, num):
```
this could be a `@staticmethod` (without the `cls` argument)

glb replied:
👍

alexcasalboni commented:
I think the implementation is flexible and robust enough. The only thing I find confusing is the renaming (which is both out-of-scope and not-backward-compatible). I'd be ok with your implementation, if we can keep the SSM naming.

Given the current status of the product, stores will be an internal detail that we don't really need to document for now. As long as we support environment variables without changing the interface, I think we are good :)

glb commented:
I've addressed the comments on this PR; will need to update with the results of the changes on other in-flight PRs (most specifically the injection method). Let me know what you think.

benkehoe commented:
This seems like overkill to me. I get the need for a swappable backing store, to support local execution. But a chained store doesn't seem like it adds value, just extra code.
Additionally, for a local store, is the environment the store to use? The names are incompatible. Why not load properly SSM-like parameters from a JSON document, either a file or a string stored in an env var?

glb commented:
@benkehoe what would you suggest for a mechanism to switch backing stores? I don't think that someone should have to make a code change to switch and suggested using an environment variable.

Can you say more about what you'd expect to see in the JSON document? Strictly name-value pairs? Learning more about SSM parameter hierarchy (https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-paramstore-su-organize.html) I'd start to expect something that can support multi-level access. 

I'd have to assume that someone will pipe up and say "YAML please" as well.

Related: will need to try a `StringList` parameter to see how that's getting handled currently; I'd assume that a JSON backend would ideally represent these as proper lists.

Meta: I'm back to work this week so will be working on this in off hours; expect a corresponding reduction in responsiveness :/

benkehoe commented:
I would think it'd look something like there's a factory function to get the default store; it checks an env var for determining whether to return the SSM or JSON/YAML-backed store (which itself uses env vars for configuration).

As for what that document might look like, it seems like it'd be useful if someone was working on a format for defining what you want the contents of SSM to be... 😉  https://github.com/iRobotCorporation/ssm-ctl/blob/master/ssm.yaml

glb commented:
Having now looked at iRobotCorporation/ssm-ctl and the suggested file format, I think I'm officially out of my depth Python-wise to resolve this issue, and with limited time available I'm unlikely to be able to learn quickly enough to close this in a timely fashion.

Please feel free to take my changes as a starting point for someone else to carry on, or as seems more likely, close this PR unmerged as an attempt that didn't meet the requirements -- the `os.environ` idea clearly doesn't come close to being able to map to real SSM parameter names.
